scripts: align sram .text section when userspace and code reloation enabled

wentongwu opened the pull request:
1) Code relocation causes comipler randomly generates _stub that will
    cause different memory layout in the magic building process for user
    mode. And the different memory layout will generate wrong hash table
    in priv_stacks_hash.c and it will cause user mode doesn't work.
    Align sram .text section when userspace and code relocation enabled
    to generate same .rodata, .data and .bss sections in sram region.
    User can adjust Kconfig RELOCATION_SRAM_TEXT_ALIGN_SIZE to make less
    wasting memory.

    2) make mpu align inside sections instead of outside to
    avoid overlap.

    3) fix spell typo in target_relocation.cmake.

Fixes: #17038.

wentongwu commented:
@andrewboie @ioannisg @nashif could you please take a look this change?Thanks

andrewboie commented:
I think I understand this patch, can you describe in a bit more detail how the memory map is shifting before this fix was applied?

wentongwu commented:
> I think I understand this patch, can you describe in a bit more detail how the memory map is shifting before this fix was applied?

@andrewboie please see issue #17038 .

wentongwu commented:
push again to solve conflict with master.

andrewboie commented:
OK, thanks for the further details.
If I understand correctly, the issue is that the ".text.sedi_pm_pmu2nvic_isr.__stub" gets added to the text section in zephyr.elf, but not zephyr_prebuilt.elf. Something is causing this to be pulled into only the final binary. At this time it is not known what is causing this to happen. This patch is a workaround to reserve some room in the prebuilt binary for text area expansion, so that the memory addresses don't change in between the two builds.

wentongwu commented:
shippable fail not related to my change
1. up_squared/samples/boards/up_squared/gpio_counter/sample.board.up_squared.gpio_counter   failure: build_error
2. qemu_xtensa/tests/kernel/sched/preempt/kernel.sched.preempt   failure: timeout
3. up_squared/samples/synchronization/sample.synchronization failure: build_error

Re-submit.

wentongwu commented:
> OK, thanks for the further details.
> If I understand correctly, the issue is that the ".text.sedi_pm_pmu2nvic_isr.__stub" gets added to the text section in zephyr.elf, but not zephyr_prebuilt.elf. Something is causing this to be pulled into only the final binary. At this time it is not known what is causing this to happen. This patch is a workaround to reserve some room in the prebuilt binary for text area expansion, so that the memory addresses don't change in between the two builds.

yes, that's workaround for it so far. 
could you please take a look this change again, there is no code change for the re-submit, just to pass the shippable, and someone is waiting this change. Thanks a lot. @andrewboie

andrewboie commented:
What steps have you taken to isolate why the extra data in the text section is being generated for zephyr.elf and not zephyr_prebuilt.elf? Where are these symbols coming from? Are they somehow being pulled in due to the gperf code we add in this phase?

> someone is waiting for this change

If someone is waiting on this patch urgently they can cherry-pick it.

andrewboie commented on Kconfig.zephyr:
```diff
@@ -82,6 +82,15 @@ config CODE_DATA_RELOCATION
+	default 80000
+	depends on ARCH_HAS_USERSPACE && CODE_DATA_RELOCATION
+	help
+	  Align sram text section when userspace and code relocation
```
This help text really doesn't explain what is going on. You need to expand on this.

andrewboie commented on Kconfig.zephyr:
```diff
@@ -82,6 +82,15 @@ config CODE_DATA_RELOCATION
+	help
+	  Align sram text section when userspace and code relocation
+	  enabled to generate same .rodata, .data and .bss sections
+	  in sram region for the magic building process.
```
"magic building process" is NOT a technical term

andrewboie commented on Kconfig.zephyr:
```diff
@@ -82,6 +82,15 @@ config CODE_DATA_RELOCATION
 
+config RELOCATION_SRAM_TEXT_ALIGN_SIZE
+	int "code relocation sram text section align size"
+	default 80000
```
What does this mean? We are by default reserving 80 Kilobytes of space? Why so much, given that the offset in the binary was only 8 bytes? The help text needs to clearly explain the semantics on the value we set here.

andrewboie commented on cmake/linker/ld/target_relocation.cmake:
```diff
@@ -3,28 +3,31 @@
 macro(toolchain_ld_relocation)
-  set(MEM_RELOCATAION_LD   "${PROJECT_BINARY_DIR}/include/generated/linker_relocate.ld")
-  set(MEM_RELOCATAION_SRAM_DATA_LD
+  set(MEM_RELOCATION_LD   "${PROJECT_BINARY_DIR}/include/generated/linker_relocate.ld")
```
Did you make some whitespace changes or something? I can't tell the difference between the before and after, it looks like the code is unchanged.

andrewboie replied:
Disregard, I see it now

andrewboie commented on cmake/linker/ld/target_relocation.cmake:
```diff
@@ -3,28 +3,31 @@
     )
 
-  add_library(code_relocation_source_lib  STATIC ${MEM_RELOCATAION_CODE})
+  add_library(code_relocation_source_lib  STATIC ${MEM_RELOCATION_CODE})
```
same as above, why is there even a diff?

wentongwu commented:
> What steps have you taken to isolate why the extra data in the text section is being generated for zephyr.elf and not zephyr_prebuilt.elf? Where are these symbols coming from? Are they somehow being pulled in due to the gperf code we add in this phase?

@andrewboie @jocelyn-li  @nashif One method I used is to put *stub to one separate section and do small align, but it doesn't work. Another idea is trying to put the .text section behind .data section, so the extra code will not impact .data section, I didn't do this one in afraid of broken other boards. And finally it's the workaround method, it works well, but now I know it doesn't fit the opensource spirit, I'm sorry for that. 

Look at it again today, I find the extra stubs are comprised of veneer functions, that's because the distance between calling code and the called function is too far(because of relocation, also there are other reasons to cause the far distance), the linker will automatically generate and insert veneer functions. Now I'm trying to figure out the solution.  And now I want to close this PR.

wentongwu committed:
scripts: align sram .text section when code reloation enabled

Code relocation causes comipler randomly generates _stub that will
cause different memory layout in the magic building process for user
mode. And the different memory layout will generate wrong hash table
in priv_stacks_hash.c and it will cause user mode doesn't work.

Align sram .text section when userspace and code relocation enabled
to generate same .rodata, .data and .bss sections in sram region.
User can adjust Kconfig RELOCATION_SRAM_TEXT_ALIGN_SIZE to make less
wasting memory.

Fixes: #17038.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
scripts: make mpu align in sections.

make mpu align inside sections instead of outside to
avoid overlap.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

wentongwu committed:
target_relocation.cmake: fix spell typo

fix spell typo in target_relocation.cmake.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>
